source/s3_request_messages.c

```diff
@@ -416,111 +416,103 @@ struct aws_http_message *aws_s3_upload_part_copy_message_new(
     }
 
-    const struct aws_byte_cursor head_operation = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("HEAD");
-    if (aws_http_message_set_request_method(message, head_operation)) {
-        goto error_clean_up;
+    struct aws_byte_cursor source_header;
+    const struct aws_byte_cursor copy_source_header = AWS_BYTE_CUR_INIT_FROM_STRING_LITERAL("x-amz-copy-source");
+    if (aws_http_headers_get(headers, copy_source_header, &source_header) != AWS_OP_SUCCESS) {
+        AWS_LOGF_ERROR(AWS_LS_S3_GENERAL, "CopyRequest is missing the x-amz-copy-source header");
+        return NULL;
     }
-
-    char destination_path[1024];
-    snprintf(destination_path, sizeof(destination_path), "/%.*s", (int)source_key.len, source_key.ptr);
-    /* TODO: url encode */
-
-    if (aws_http_message_set_request_path(message, aws_byte_cursor_from_c_str(destination_path))) {
-        goto error_clean_up;
+    struct aws_byte_cursor host;
+    if (aws_http_headers_get(headers, g_host_header_name, &host) != AWS_OP_SUCCESS) {
+        AWS_LOGF_ERROR(AWS_LS_S3_GENERAL, "CopyRequest is missing the Host header");
+        return NULL;
     }
 
-    char host_header_value[1024];
-    /* TODO: Fix the hard-coded host name. */
-    snprintf(
-        host_header_value,
-        sizeof(host_header_value),
-        "%.*s.s3.us-west-2.amazonaws.com",
-        (int)source_bucket.len,
-        source_bucket.ptr);
-    struct aws_http_header host_header = {
-        .name = g_host_header_name,
-        .value = aws_byte_cursor_from_c_str(host_header_value),
-    };
-    aws_http_message_add_header(message, host_header);
+    struct aws_byte_cursor request_path = source_header;
 
-    aws_http_message_set_body_stream(message, NULL);
+    /* Skip optional leading slash. */
+    if (aws_byte_cursor_starts_with(&request_path, &s_slash_char)) {
+        aws_byte_cursor_advance(&request_path, 1);
+    }
 
-    return message;
+    /* From this point forward, the format is {bucket}/{key} - split
+    components.*/
 
-error_clean_up:
+    struct aws_byte_cursor source_bucket = {0};
 
-    if (message != NULL) {
-        aws_http_message_release(message);
-        message = NULL;
+    if (aws_byte_cursor_next_split(&request_path, '/', &source_bucket)) {
+        aws_byte_cursor_advance(&request_path, source_bucket.len);
     }
 
-    return NULL;
-}
-
-/* Creates a HEAD GetObject sub-request to get the size of the source object of a Copy meta request. */
-struct aws_http_message *aws_s3_get_source_object_size_message_new(
-    struct aws_allocator *allocator,
-    struct aws_http_message *base_message) {
-    AWS_PRECONDITION(allocator);
+    if (source_bucket.len == 0 || request_path.len == 0) {
+        AWS_LOGF_ERROR(AWS_LS_S3_GENERAL, "CopyRequest x-amz-copy-source header not in bucket/key format");
+        goto error_cleanup;
+    }
 
-    struct aws_http_message *message = NULL;
+    if (aws_byte_buf_init_copy_from_cursor(&head_object_host_header, allocator, source_bucket)) {
```

Comment:
we could, but code looks slightly uglier since we need to break init into init + append. Personally, i prefer this variant slightly more even if its a little less efficient